### ISSUE TITLE
Cannot get example to work.

### ISSUE THREAD

**Issue by AlanGRutter** [NapalmCodes/Aspire.Hosting.Krakend#31]
I was hoping this was going to be an easy package to install and run but I appear to having issues. I have followed your example and created a krakend.json file. When I start my Aspire Host, the KrakenD container fails to start correctly with the following error.

 ERROR parsing the configuration file: failed to copy source file /tmp/KrakenD_parsed_config_template_852974428 to /etc/krakend/result.json: open /etc/krakend/result.json: permission denied

I am running on Ubuntu if that makes any difference.
[krakend.json](https://github.com/user-attachments/files/17640597/krakend.json)

Would greatly appreciate some feedback. I also assume that the host config in your example is used to dynamically get the docker internal address for a service.

In my instance, I want to run Krakend with a Dapr sidecar because I want to communicate with my backend services through the sidecar. By adding  

```
"modifier/martian": 
{
    "header.Append": {
        "scope": ["request"],
        "name": "dapr-app-id",
        "value": "<dapr app id of service>"
    }
}
```
and since all Dapr sidecars run on port 3500 I can either set that as the default host or use it as follows
```
{
  "endpoint": "/user",
  "method": "GET",
  "backend": [
    {
      "url_pattern": "https://localhost:3500/health",
      "method": "GET",
      "encoding": "string"
    }
  ]
},
```

**Comment by NapalmCodes** [comment]
I have not attempted to run on Linux but I am wondering if the issue is a file permissions problem in mounting the volume with the configuration to docker? I could be wrong but Docker Desktop sets this automatically to 0777 for volume mounts from Windows. Small suggestion, turn off `FC_ENABLE` setting it to `0` and maybe make a vanilla configuration (i.e.: without dynamic configuration) file to see if that works for troubleshooting purposes (assuming adjusting the permissions doesn't take care of it first)?

**Comment by AlanGRutter** [comment]
Do you have a simple config that you know works which you could attach?

Sent from my phone.

On 7/11/2024, at 1:10 PM, Shawn Vause ***@***.***> wrote:

﻿

I have not attempted to run on Linux but I am wondering if the issue is a file permissions problem in mounting the volume with the configuration to docker? I could be wrong but Docker Desktop sets this automatically to 0777 for volume mounts from Windows. Small suggestion, turn off FC_ENABLE setting it to 0 and maybe make a vanilla configuration (i.e.: without dynamic configuration) file to see if that works for troubleshooting purposes (assuming adjusting the permissions doesn't take care of it first)?

**Comment by NapalmCodes** [comment]
I mean an example is in the example here: https://github.com/NapalmCodes/Aspire.Hosting.Krakend/blob/main/examples/BasicGateway/BasicApiGateway/BasicApiGateway.AppHost/config/krakend/result.json

Just update your host section for your services IP. If you don't want to worry about OTEL for now you can yank lines 29-83. That is about as basic as it gets.

**Comment by AlanGRutter** [comment]
I attached a krakend.json file in the original message. Are you able to verify if it is good or has errors - I couldn't see any.

**Comment by NapalmCodes** [comment]
It is invalid:

Run the following where C:/temp is whatever dir has your JSON.
```
docker run --rm -it -v C:/temp:/etc/krakend devopsfaith/krakend sh
/etc/krakend # krakend check -tlc krakend.json
Parsing configuration file: krakend.json
ERROR linting the configuration file:   jsonschema: '/tls' does not validate with https://www.krakend.io/schema/v2.4/krakend.json#/properties/tls/$ref/additionalProperties: additionalProperties 'keys' not allowed
```
 Pro tip: A validator is built into the docker image: https://www.krakend.io/docs/configuration/check/

**Comment by NapalmCodes** [comment]
FYI I did also see a similar issue today. Try deleting `results.json` out of the folder you are placing it in (assuming you are using `FC_OUT`) and rebuild. I was seeing a permission problem with the resultant file today as I was working on feature updates forthcoming.

**Comment by AlanGRutter** [comment]
I have got a krakend.json file that yields 'Syntax OK' when I check it. I don't have any results.json file in my config directory and when I run the Aspire Host, the Krakend container exits immediately with a permission error copying the file to /etc/krakend/

**Comment by NapalmCodes** [comment]
Hmm. Did you ever see the result file locally? When I saw a similar issue yesterday the problem resolved when I deleted that result file between runs. Was leading me to consider a pre build action to delete the old file between runs.

**Comment by AlanGRutter** [comment]
No - I only got a results file when I manually ran the check and output command. When using the AddKrakend command, I've never seen a local results file. So having watched a video on Aspire, I can see where the services__ names come from - I assume  the idea is that when you create the results file, it substitutes the correct values. I tried manually setting the hosts in the config file to the contents of those environment varaibles (e.g."http://host.containers.internal:5067") but that doesn't see to be working either.

**Comment by NapalmCodes** [comment]
Strange, but yes flexible config uses Go (golang) templating to substitute the values in. I assume you looked at this: https://www.krakend.io/docs/configuration/flexible-config/

This feels like a difference in behavior using Docker Desktop vs native Docker if the config file you are providing is valid as the linter suggests. I'm not sure it's an Aspire issue per se since it works without issue on Windows/Docker Desktop (admittedly I haven't tried podman). Are you using native docker I presume? I wonder if you still need docker desktop on Linux to be supported. Documented in link below by Microsoft. It could be they are assuming the appropriate permissions are setup by DD/Podman as they are dev tools like Aspire itself is.

I would probably need to spin up a Linux environment myself to figure this out. I would say wsl2 but it's using the same docker desktop vm as windows does.

Did you try a static json file with the values for your api hardcoded (i.e.: produce the results file yourself and just mount that in with fc and the results path off/not provided)? Does the gateway serve that traffic to your api as expected?

Another thought let's take the local folder out completely (delete it) and let the container create it from scratch with permissions (per discussion in the issue linked below) from what you pass in the parameter for the path). Then put the config file in that folder and restart?

Check out this issue it seems similar to me and has a few things to look at (yes I understand he is using Rancher but this is a permissions issue for certain):  https://github.com/dotnet/aspire/issues/4956

Sorry to throw all this at you, I just am not experiencing these issues myself and don't have an environment handy. I wouldn't be afraid to ask the Aspire team too potentially. If this is a Linux native docker thing I am sure someone ran into it before you. If I think of anything I'll add it but please report back it may help others! I have been unable to find anything explicit indicating aspire is supported for native docker on Linux.

**Comment by AlanGRutter** [comment]
I'm using Podman and Podman Desktop on a Ubuntu development machine which is an Amazon Workspace. I don't have docker installed at all on this machine. I will doing some more testing with static files but it would be helpful if you could stand up a Ubuntu instance on Amazon Workspace (free tier available) and see if you can replicate. I'm happy to work with you - give me a shout or a private message if you want to do something together. I'm in New Zealand.

**Comment by NapalmCodes** [comment]
Curious if you had any luck? Recent changes suggested using /tmp for results and the example has been updated in the latest release. I have talked with some people and the assumption is SEL Linux is in use here and you are running into host permission issues given functionality on MacOS and Windows is fully functional.

**Comment by AlanGRutter** [comment]
I did upgrade to the latest version but I didn’t have time to look at the changes. I will hopefully be able to check soon.

Sent from my phone.

On 20/11/2024, at 5:31 PM, Shawn Vause ***@***.***> wrote:

﻿

Curious if you had any luck? Recent changes suggested using /tmp for results and the example has been updated in the latest release. I have talked with some people and the assumption is SEL Linux is in use here and you are running into host permission issues given functionality on MacOS and Windows is fully functional.
